Improvements to refresh token rotation

graziang opened the pull request:
With `Revoke Refresh Token` enabled, only the last used refresh token with its number of uses was saved. In this way it was impossible to use refresh tokens with the same client on multiple browser tabs.

To address this, when a refresh token is used each refresh token is saved in the client session notes along with the number of token uses. In this way it is possible to keep track of the used tokens and detect if the maximum number of uses is reached.

In `Revoke Refresh Tokens` settings in addition to `Refresh Token Max Reuse` is now possible to enter a `Refresh Token Reuse Period` (in seconds). 

When a refresh token is used it can remain valid for a specified duration. To address this, in the client session notes is saved the last refresh timestamp for each refresh token.

As a possible side effect I see a large number of notes saved for each refresh token in the client session.

Closes #14122

rmartinc reviewed (changes requested):
Just a few minor comments and a big concern I have about the number of notes in the client session. Maybe I'm overthinking it, but I don't like the idea of adding new notes for each refresh token generated. There can be too many notes. So I'm proposing some solutions for the problem if you also see it problematic.

graziang commented on services/src/main/java/org/keycloak/protocol/oidc/TokenManager.java:
```diff
@@ -456,28 +457,15 @@ private void validateTokenReuseForRefresh(KeycloakSession session, RealmModel re
-            throw new OAuthErrorException(OAuthErrorException.INVALID_GRANT, "Stale token");
-        }
+    private void validateTokenReuse(RealmModel realm, AccessToken refreshToken, AuthenticatedClientSessionModel clientSession) throws OAuthErrorException {
+        int lastRefresh = clientSession.getRefreshTokenLastRefresh(refreshToken.getId());
```
Using `iat` allows the refresh token to be used only for X seconds (Reuse Period) after `iat`, but the intent is to allow the refresh token to be reused immediately after its refresh for X seconds (Reuse Period). 
In other words, you should start counting the reuse period from the moment of the refresh and not from the moment of issuance (`iat`) of the refresh token. 
An improvement could be to save the last refresh only when the maximum reuse limit is reached to avoid potential infinite reuse of the same refresh token in case it is refreshed during the "reuse period" just after refresh.

rmartinc replied:
Ah! OK, so it's the time of the new refresh, not the previous one. Understood. Please resolve this conversation (I cannot do it, I suppose I have no permissions).

graziang commented on services/src/main/java/org/keycloak/protocol/oidc/TokenManager.java:
```diff
@@ -436,14 +436,15 @@ public AccessTokenResponseBuilder refreshAccessToken(KeycloakSession session, Ur
-                clientSession.setCurrentRefreshTokenUseCount(currentCount + 1);
+                validateTokenReuse(realm, refreshToken, clientSession);
+                int currentCount = clientSession.getRefreshTokenUseCount(refreshToken.getId());
+                clientSession.setRefreshTokenUseCount(refreshToken.getId(), currentCount + 1);
```
I have the same concerns as you about the tons of client notes that can be created, I'm not sure if saving something like max usage in the refresh token will help since you still need to count the number of uses for each token. As you mentioned, cleaning up notes with expired tokens can help a bit, but it's not a solution for offline tokens.

Maybe a solution could be to save something like the `tab_id` in the refresh token, this way you can collect refresh tokens per tab and only keep in the client notes the last one used for each tab. WDYT?

mposolda replied:
+1 for using `tabId` to track tokens. Moreover, I think that we should make sure that just one refresh token is valid per tabId. So for instance when refreshToken from `tab1` is refreshed and new refresh token `rt2` is generated for this tab, then after successfully refresh with using `rt2`, the previous `rt1` should not be valid anymore.

Otherwise, it would be possible that if some attacker manage to steal refresh token `rt1` and manage to successfully refresh it and generate refresh token `rt2-evil` and at the same time, the same refresh token `rt1` is used by correct user to generate another refresh token `rt2-good`, then both refresh tokens `rt2-evil` and `rt2-good` would be valid and would be possible to refresh them both successfully many time without detection that attacker successfully managed to steal the refresh token. This is exactly what we want to prevent here as IMO the main point of "revoke refresh token" feature is to prevent the case when someone manage to steal refresh token... We should make sure that there is only single "branch" valid for every `tabId` and hence there is still need to track somehow the refresh token per tab-id in clientSession though...

This is also aligned with the tooltips in admin console UI on fields `revokeRefreshTokenHelp` and `refreshTokenMaxReuseHelp`, which both mention that token is immediatelly revoked after it is successfully used.

So IMO it can be good if new methods on `AuthenticatedClientSessionModel` consider `tabId` rather than `refreshTokenId` and also that there is still `String getCurrentRefreshToken(String tabId)` . Also it looks that we may need to add something like `tabId` claim on refreshToken to make sure that refresh is successful just with refresh token, which matches `current refresh token` for the tab or where timestamp is equal to `iat` (so we may need separate note for last timestamp for the tab as well as field `timestamp` on `AuthenticatedClientSessionModel` cannot be used IMO due refreshes from many tabs).

Also using `tabId` should handle maximum limit for notes automatically as there is limit for maximum count of "Tabs" already per authentication session, which is 300 by default ( `InfinispanAuthenticationSessionProviderFactory.DEFAULT_AUTH_SESSIONS_LIMIT` ).

mposolda reviewed (commented):
@graziang @rmartinc Thanks for the PR and review!

Added comment inline regarding the use of "tabId" to track tokens.

My other concern is, that we should make sure that it is still possible to refresh with the old offline token after migration from previous version. In other words, offline token generated in Keycloak 24, can be still used in Keycloak 25 (or whatever version we managed to have this PR in). It seems to me that this won't work right now based on the changes in the PR (but maybe I am wrong...).
There is `AbstractMigrationTest.migration24_xTest()` . I wonder if we want similar method also for the case with "revoke refresh token" enabled? At the same time, not sure how much effort and changes needed in `MigrationTest` for this, so maybe testing this manually is ok?  WDYT?

Finally, I am not sure if we need new switch `refreshTokenReusePeriod` on the realm and the related functionality. Any particular use-case why the switch would be useful to address any issues described in https://github.com/keycloak/keycloak/issues/14122 ?

graziang committed:
Improvements to refresh token rotation
Closes #14122

Signed-off-by: Giuseppe Graziano <g.graziano94@gmail.com>

graziang commented:
@rmartinc @mposolda Thanks for the review!

I've updated the PR:

- Introduced a claim in the refresh token called `reuse_id`, generated with a new UUID each time a refresh token is generated with a grant different from the refresh token grant. When a refresh token request occurs, the `reuse_id` is copied from the old one to the new refresh token. The `tab_id` was not easy to obtain during token generation for the code grant, but the behavior should be the same.

- The `reuse_id` is used as a key to store in the client session the last used refresh token id, the last refreshed timestamp, and the count of refresh token usages. So for each `reuse_id` (similar to `tab_id`), you have this triple of notes.

- The checks remain the same but are based on the triple of client notes obtained using the `reuse_id`.

- An old refresh_token (without the `reuse_id` claim) can still be used. This means that, for instance, offline tokens generated in Keycloak 24 can still be used with the Keycloak version containing this PR, but existing refresh token usage counts would be reset. Can this work? 
This change allows us to remove the old methods from the `AuthenticatedClientSessionModel` and the related code. Alternatively, we should keep them and use them in case of a missing `reuse_id` in the refresh token for backward compatibility.

- Removed the `refreshTokenReusePeriod` feature. This feature was discussed in the comments on the issue. A potentially useful use case might be when users with unstable network connections don't receive new tokens on the client app, and as a result the client may need to reuse the same refresh token to request new tokens. However, this can also be solved with `Refresh Token Max Reuse` = 1, without introducing a "reuse period." WDYT?

mposolda commented on services/src/main/java/org/keycloak/protocol/oidc/TokenManager.java:
```diff
@@ -1104,10 +1106,13 @@ public AccessTokenResponseBuilder generateRefreshToken() {
         }
 
-        public AccessTokenResponseBuilder generateRefreshToken(String scope, AuthenticatedClientSessionModel clientSession) {
+        public AccessTokenResponseBuilder generateRefreshToken(String scope, AuthenticatedClientSessionModel clientSession, Object reuseId) {
```
```suggestion
        public AccessTokenResponseBuilder generateRefreshToken(RefreshToken oldRefreshToken, AuthenticatedClientSessionModel clientSession) {
```

Maybe a nitpick, but is it possible to update signature of this method to something like the above? This can make it clearer that it is supposed to be used to generate refresh token from the old refresh token (both `scope` and `reuseId` could be retrieved from it).

mposolda commented on services/src/main/java/org/keycloak/protocol/oidc/TokenManager.java:
```diff
@@ -442,8 +443,10 @@ private void validateTokenReuseForRefresh(KeycloakSession session, RealmModel re
+                String key = Optional.ofNullable(refreshToken.getOtherClaims().get(Constants.REUSE_ID)).map(String::valueOf).orElse("");
+                int currentCount = clientSession.getRefreshTokenUseCount(key);
+                clientSession.setRefreshTokenUseCount(key, currentCount + 1);
+                clientSession.setRefreshTokenLastRefresh(key, Time.currentTime());
```
I can think of the corner case that `Time.currentTime()`, which is generated on this line maybe be slightly different from the time in the `iat` claim of the refresh token. It seems there can be some small difference (like EG. 1 second) among them. I can imagine this can cause some issues in the corner scenarios...

I've noticed that similar issue already exists on this line https://github.com/keycloak/keycloak/blob/main/services/src/main/java/org/keycloak/protocol/oidc/TokenManager.java#L1127 . Is it possible to update this to use something like `int currentTime = refreshToken.getISsuedAt();` to make sure that it is really same time in the refreshToken and clientSession/userSession and also make sure that `clientSession.setRefreshTokenLastRefresh` is updated with this timestamp as well?

Maybe this is a nitpick, but from the past experience, the similar things can cause troubles in production or security issues or flaky tests in our CI (and maybe this one will fix some flaky tests as a side-effect :-)

mposolda reviewed (commented):
@graziang Thanks for the updates! To me, it looks good and just added some nitpick comments....

Some more feedback to your questions:



> An old refresh_token (without the reuse_id claim) can still be used. This means that, for instance, offline tokens generated in Keycloak 24 can still be used with the Keycloak version containing this PR, but existing refresh token usage counts would be reset. Can this work?

IMO this is fine.

> This change allows us to remove the old methods from the AuthenticatedClientSessionModel and the related code. Alternatively, we should keep them and use them in case of a missing reuse_id in the refresh token for backward compatibility.

I afraid we need to keep the methods due it is problematic to just remove methods from `server-spi` module as people may rely on those classes in their providers (it is supposed to be Keycloak public SPI) and in theory, this can cause compilation issues. But I would keep the "logic" as you did it and not use those methods from any Keycloak code. So we just keep those methods in order to not break compilation, but we won't use them... Added also one inline comment about this.


> Removed the refreshTokenReusePeriod feature. This feature was discussed in the comments on the issue. A potentially useful use case might be when users with unstable network connections don't receive new tokens on the client app, and as a result the client may need to reuse the same refresh token to request new tokens. However, this can also be solved with Refresh Token Max Reuse = 1, without introducing a "reuse period." WDYT?

+1 to not adding `refreshTokenReusePeriod`  this. As you mentioned, I think that in case of unstable network, we are still fine due `Refresh Token Max Reuse` . Some more context in the old Keycloak JIRA https://issues.redhat.com/browse/KEYCLOAK-3303 (related commit was https://github.com/keycloak/keycloak/commit/134daeac ).

graziang committed:
Follow-up improvements to refresh token rotation

Closes #14122

Signed-off-by: Giuseppe Graziano <g.graziano94@gmail.com>

mposolda reviewed (dismissed):
@graziang Thanks!

rmartinc commented on model/storage-private/src/main/java/org/keycloak/models/session/PersistentAuthenticatedClientSessionAdapter.java:
```diff
@@ -302,10 +282,6 @@ protected static class PersistentClientSessionData {
         private Set<String> protocolMappers;
         @JsonProperty("roles")
         private Set<String> roles;
-        @JsonProperty("currentRefreshToken")
```
Take care with this. It seems that this is serialized inside the database and if properties are removed an exception is thrown when an offline session is restored from database. Check the comment just before this point. I just added this line:

```java
    @JsonIgnoreProperties(ignoreUnknown = true)
    protected static class PersistentClientSessionData {
```

It seems that now in quarkus it works OK.

mposolda replied:
@rmartinc Good catch! +1 for adding the test.

There is already `AbstractMigrationTest.testOfflineTokenLogin()`, but looks that this one is not tested with "revoke refresh token" enabled? Maybe we can just update existing realm(s) JSON to use "revoke refresh token" though? At least for the last migration (probably `migration-realm-19.0.3.json`) to make sure it is tested in GH actions.

graziang replied:
Thanks @rmartinc  and @mposolda ! Running migration tests with `revokeRefreshToken` enabled in `migration-realm-24.0.4.json`, I was unable to reproduce the issue. After some investigation, it seems that `currentRefreshToken` and `currentRefreshTokenUseCount` of `PersistentClientSessionData` are not being saved in the database. To be sure I have added `@JsonIgnoreProperties(ignoreUnknown = true)` to `PersistentClientSessionData` and enabled `revokeRefreshToken` in `migration-realm-24.0.4.json`.

rmartinc reviewed (changes requested):
Thanks @graziang! LGTM in general. But I did a quick test of creating an offline session with Revoke refresh token activated to 3 and then changed to the new code and serialization was broken when retrieving from database. I needed to add `@JsonIgnoreProperties(ignoreUnknown = true)` to the `PersistentClientSessionData`.

@graziang @mposolda Maybe we need a test for migration? WDYT?

The rest worked for me OK.

andymunro reviewed (approved):
Looks good, @mposolda

graziang committed:
Migration test for offline tokens

Closes #14122

Signed-off-by: Giuseppe Graziano <g.graziano94@gmail.com>

rmartinc reviewed (approved):
Thanks @graziang! Approved!

mposolda reviewed (approved):
@graziang @rmartinc @andymunro Thanks for all the work on this PR!

martyfitzer-ids commented:
I am looking to utilize this refresh token fix in my application. Does anyone have an idea on when this will be released now that it is merged?

graziang commented:
> I am looking to utilize this refresh token fix in my application. Does anyone have an idea on when this will be released now that it is merged?

Should be released soon with Keycloak 25

ahus1 commented:
@graziang, @mposolda, @rmartinc - in our team that created the persistent sessions we were a bit surprised that this PR changed the  serialization of sessions in AuthenticatedClientSessionEntity. In general, those changes usually break serialization if they are not versioned. If someone is using an external Infinispan (like in multi-site setups), one would need to clear all sessions there manually during an upgrade to avoid reading corrupted data. 

We're a bit unlucky here as the serializer of AuthenticatedClientSessionEntity doesn't write a version number as the first field as other serializers do this.

For this change, it seems we have been lucky still: The fields removed are from the end of the serialization stream, and we are never writing two clients sessions in one stream. I did a manual migration test and seems that it worked.

I'm sharing this here to to raise awareness for future changes in the KC25 and previous releases. 

For KC26, we're planning to move to Protostream marshalling, plus additional safeguards that serialization is always backwards compatible, see #29474 for the first step. 

cc: @mhajas, @pruivo

martyfitzer-ids commented:
I am looking to utilize this fix in my application, but it is a little unclear to me how to do so. Can someone please summarize what this PR changes, what kind of setting changes I need to make (if any), and what the expected behavior should be once this is implemented? Thanks!
